File: packages/python/elizaos/features/advanced_capabilities/actions/image_generation.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import IMAGE_GENERATION_TEMPLATE
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_action_spec("GENERATE_IMAGE")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    spec_examples = _spec.get("examples", [])
    if not isinstance(spec_examples, list):
        return []
    result: list[list[ActionExample]] = []
    for example in spec_examples:
        if not isinstance(example, list):
            continue
        row: list[ActionExample] = []
        for msg in example:
            if not isinstance(msg, dict):
                continue
            content = msg.get("content", {})
            text = ""
            actions: list[str] | None = None
            if isinstance(content, dict):
                text_val = content.get("text", "")
                text = str(text_val) if text_val else ""
                actions_val = content.get("actions")
                if isinstance(actions_val, list) and all(isinstance(a, str) for a in actions_val):
                    actions = list(actions_val)
            row.append(
                ActionExample(
                    name=str(msg.get("name", "")),
                    content=Content(text=text, actions=actions),
                )
            )
        if row:
            result.append(row)
    return result


@dataclass
class GenerateImageAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        return runtime.has_model(ModelType.IMAGE)

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if state is None:
            raise ValueError("State is required for GENERATE_IMAGE action")

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ACTION_STATE"])

        template = (
            runtime.character.templates.get("imageGenerationTemplate")
            if runtime.character.templates
            and "imageGenerationTemplate" in runtime.character.templates
            else IMAGE_GENERATION_TEMPLATE
        )
        prompt = runtime.compose_prompt(state=state, template=template)

        prompt_response = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(prompt_response)

        if parsed_xml is None:
            raise ValueError("Failed to parse XML response for image prompt")

        thought = str(parsed_xml.get("thought", ""))
        image_prompt = str(parsed_xml.get("prompt", ""))

        if not image_prompt:
            raise ValueError("No image prompt generated")

        image_result = await runtime.use_model(
            ModelType.IMAGE,
            prompt=image_prompt,
        )

        image_url: str | None = None
        if isinstance(image_result, str):
            image_url = image_result
        elif isinstance(image_result, dict):
            image_url = image_result.get("url") or image_result.get("data")

        if not image_url:
            raise ValueError("No image URL returned from generation")

        response_content = Content(
            text=f"Generated image with prompt: {image_prompt}",
            attachments=[{"type": "image", "url": image_url}],
            actions=["GENERATE_IMAGE"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Generated image: {image_prompt}",
            values={
                "success": True,
                "imageGenerated": True,
                "imageUrl": image_url,
                "imagePrompt": image_prompt,
            },
            data={
                "actionName": "GENERATE_IMAGE",
                "prompt": image_prompt,
                "thought": thought,
                "imageUrl": image_url,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


generate_image_action = Action(
    name=GenerateImageAction.name,
    similes=GenerateImageAction().similes,
    description=GenerateImageAction.description,
    validate=GenerateImageAction().validate,
    handler=GenerateImageAction().handler,
    examples=GenerateImageAction().examples,
)

File: packages/typescript/src/features/advanced-capabilities/actions/imageGeneration.ts
import { v4 } from "uuid";
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	collectKeywordTermMatches,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { imageGenerationTemplate } from "../../../prompts.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { ContentType, ModelType } from "../../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("GENERATE_IMAGE");
const IMAGE_EXTENSIONS = new Set(["png", "jpg", "jpeg", "gif", "webp", "bmp"]);
const IMAGE_STRONG_TERMS = getValidationKeywordTerms(
	"action.generateImage.strong",
	{
		includeAllLocales: true,
	},
);
const IMAGE_WEAK_TERMS = getValidationKeywordTerms(
	"action.generateImage.weak",
	{
		includeAllLocales: true,
	},
);

const getFileExtension = (url: string): string => {
	const urlPath = new URL(url).pathname;
	const lastDot = urlPath.lastIndexOf(".");
	if (lastDot === -1 || lastDot === urlPath.length - 1) {
		return "png";
	}
	const extension = urlPath.slice(lastDot + 1).toLowerCase();
	return IMAGE_EXTENSIONS.has(extension) ? extension : "png";
};

export const generateImageAction = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	validate: async (_runtime: IAgentRuntime, message: Memory) => {
		const text =
			typeof message?.content === "string"
				? message.content
				: (message?.content?.text ?? "");
		if (!text) return false;
		if (collectKeywordTermMatches([text], IMAGE_STRONG_TERMS).size > 0) {
			return true;
		}
		return collectKeywordTermMatches([text], IMAGE_WEAK_TERMS).size > 0;
	},
	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
		responses?: Memory[],
	): Promise<ActionResult> => {
		const allProviders: string[] = [];
		if (responses) {
			for (const res of responses) {
				const providers = res.content?.providers;
				if (providers && providers.length > 0) {
					allProviders.push(...providers);
				}
			}
		}

		state = await runtime.composeState(message, [
			...(allProviders ?? []),
			"RECENT_MESSAGES",
		]);

		const prompt = composePromptFromState({
			state,
			template:
				runtime.character.templates?.imageGenerationTemplate ||
				imageGenerationTemplate,
		});

		const promptResponse = await runtime.useModel(ModelType.TEXT_LARGE, {
			prompt,
			stopSequences: [],
		});

		const parsedXml = parseKeyValueXml(promptResponse);
		const promptValue = parsedXml?.prompt;

		const imagePrompt: string =
			typeof promptValue === "string"
				? promptValue
				: "Unable to generate descriptive prompt for image";

		const imageResponse = await runtime.useModel(ModelType.IMAGE, {
			prompt: imagePrompt,
		});
		const imageResults = Array.isArray(imageResponse)
			? imageResponse
			: typeof imageResponse === "string"
				? [imageResponse]
				: [];
		const firstImage = imageResults[0];
		const firstImageUrl =
			typeof firstImage === "string" ? firstImage : firstImage?.url;

		if (imageResults.length === 0 || !firstImageUrl) {
			logger.error(
				{
					src: "plugin:advanced-capabilities:action:image_generation",
					agentId: runtime.agentId,
					imagePrompt,
				},
				"Image generation failed - no valid response received",
			);
			return {
				text: "Image generation failed",
				values: {
					success: false,
					error: "IMAGE_GENERATION_FAILED",
					prompt: imagePrompt,
				},
				data: {
					actionName: "GENERATE_IMAGE",
					prompt: imagePrompt,
					rawResponse: imageResults.map((image) => ({
						url: typeof image === "string" ? image : image.url,
					})),
				},
				success: false,
			};
		}

		const imageUrl = firstImageUrl;

		logger.info(
			{
				src: "plugin:advanced-capabilities:action:image_generation",
				agentId: runtime.agentId,
				imageUrl,
			},
			"Received image URL",
		);

		const extension = getFileExtension(imageUrl);
		const timestamp = new Date()
			.toISOString()
			.replace(/[:.]/g, "-")
			.slice(0, 19);
		const fileName = `Generated_Image_${timestamp}.${extension}`;
		const attachmentId = v4();

		const responseContent = {
			attachments: [
				{
					id: attachmentId,
					url: imageUrl,
					title: fileName,
					contentType: ContentType.IMAGE,
				},
			],
			thought: `Generated an image based on: "${imagePrompt}"`,
			actions: ["GENERATE_IMAGE"],
			text: imagePrompt,
		};

		if (callback) {
			await callback(responseContent);
		}

		return {
			text: "Generated image",
			values: {
				success: true,
				imageGenerated: true,
				imageUrl,
				prompt: imagePrompt,
			},
			data: {
				actionName: "GENERATE_IMAGE",
				imageUrl,
				prompt: imagePrompt,
			},
			success: true,
		};
	},
	examples: (spec.examples ?? []) as ActionExample[][],
} as Action;
